### PR TITLE
Зайцев Денис. 3 задача. 19 вариант. Поразрядная сортировка для целых чисел с четно-нечетным слиянием Бэтчера.

### PR DESCRIPTION
**Pull request by DZaytsev4**
MPI версия:
1. Данные распределяются по процессам.
2. Происходит локальная поразрядная сортировка.
3. Процессы начинают обмениваться данными четно-нечетным слиянием Бэтчера в зависимости от шага.
4. Данные собираются в один массив.

Seq версия:
Происходит поразрядная сортировка

### CODE REVIEW

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    zaytsev_bitwise_sort_evenodd_Batcher::TestMPITaskSequential testMpiTaskSequential(taskDataSeq);
+    ASSERT_EQ(testMpiTaskSequential.validation(), true);
+    testMpiTaskSequential.pre_processing();
+    testMpiTaskSequential.run();
+    testMpiTaskSequential.post_processing();
+
+    ASSERT_EQ(global_result, sequential_result);
+  }
+}
```
Comment (ArtemPch) [inline tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:85]: add tests, for example, with random numbers and an empty vector

Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp]: ready

**Review by ArtemPch — APPROVED** [review]

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/src/ops_mpi.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  int max_value = *std::max_element(data_.begin(), data_.end());
+  int max_bits = 0;
+  while (max_value > 0) {
+    max_value >>= 1;
+    ++max_bits;
+  }
+
+  std::vector<int> buffer(data_.size());
```
Comment (BogdanNaumov) [inline tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/src/ops_mpi.cpp:44]: в seq и mpi версии используется поразрядная сортировка, её можно вывести в отдельную функцию

Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/src/ops_mpi.cpp]: ready

**Review by BogdanNaumov — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+#include <vector>
+
+#include "mpi/zaytsev_bitwise_sort_evenodd_Batcher/include/ops_mpi.hpp"
+
+static std::vector<int> getRandomVector(int sz) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<int> vec(sz);
+  for (int i = 0; i < sz; i++) {
+    vec[i] = gen() % 100 - 50;
```
Comment (allnes) [inline tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:16]: please parametrize this constants in tests

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  testMpiTaskParallel.pre_processing();
+  testMpiTaskParallel.run();
+  testMpiTaskParallel.post_processing();
+
+  if (world.rank() == 0) {
+    ASSERT_TRUE(global_result.empty());
+  }
+}
+
+TEST(zaytsev_bitwise_sort_evenodd_Batcher, RandomVector) {
```
Comment (allnes) [inline tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:121]: please add tests with 2^n, 3^n, prime values

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    ASSERT_EQ(testMpiTaskSequential.validation(), true);
+    testMpiTaskSequential.pre_processing();
+    testMpiTaskSequential.run();
+    testMpiTaskSequential.post_processing();
+
+    ASSERT_EQ(global_result, sequential_result);
+  }
+}
+
+TEST(zaytsev_bitwise_sort_evenodd_Batcher, EmptyVector) {
```
Comment (allnes) [inline tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:97]: please add test with reverse sort case

Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:125]: ready

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp (hunk 4 above)
Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:149]: ready

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp (hunk 5 above)
Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:125]: ready

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp (hunk 4 above)
Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:149]: ready

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp (hunk 3 above)
Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp]: ready

File: tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp (hunk 5 above)
Comment (DZaytsev4) [reply tasks/mpi/zaytsev_bitwise_sort_evenodd_Batcher/func_tests/main.cpp:125]: ready
